File: tests/plink_parsing/plink_test.c
#include "bayesrco_io.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    IOConfig ioconfig;
    GenomicData gdata;

    memset(&ioconfig, 0, sizeof(IOConfig));
    memset(&gdata, 0, sizeof(GenomicData));
    
    strcpy(ioconfig.geno_file_path, "test_data.bed");
    strcpy(ioconfig.pheno_file_path, "test_data.fam");
    ioconfig.trait_column_index = 1;
    ioconfig.mcmc = true;
    io_get_size(&ioconfig, &gdata);
    io_load_phenotypes(&ioconfig, &gdata);
    io_load_genotypes(&ioconfig, &gdata);

    printf("Phenotypes (why):\n");
    for (int i = 0; i < gdata.num_individuals; i++) {
        if (gdata.phenotypes[i] == MISSING_VALUE) {
            printf(" NA\n");
        } else {
            printf("%20.6f\n", gdata.phenotypes[i]);
        }
    }

    printf("Genotypes (X):\n");
    for (int i = 0; i < gdata.num_phenotyped_individuals; i++) {
        for (int j = 0; j < gdata.num_loci; j++) {
            printf("%5.1f\n", gdata.genotypes[j * gdata.num_phenotyped_individuals + i]);
        }
    }

    io_cleanup(&ioconfig, &gdata, NULL, NULL);

    return 0;
}

File: tests/plink_parsing/plink_test.f90
program plink_test
    use mod_defs, only: dp, MISSING_VALUE
    use mod_data
    use mod_io
    implicit none

    integer :: i, j

    config%inprefix = "test_data"
    config%phenfil = "test_data.fam"
    config%bimfil = "test_data.bim"
    config%genfil = "test_data.bed"
    config%trait_pos = 1
    config%mcmc = .true.

    call get_size(config, gdata)
    call load_phenos_plink(config, gdata)
    
    ! Allocate X
    gdata%nt = count(gdata%trains == 0)
    allocate(gdata%X(gdata%nt, gdata%nloci))
    
    call load_snp_binary(config, gdata)

    print *, "Phenotypes (why):"
    do i = 1, gdata%nind
        if (gdata%why(i) == MISSING_VALUE) then
            print *, "NA"
        else
            print "(F20.6)", gdata%why(i)
        end if
    end do

    print *, "Genotypes (X):"
    do i = 1, gdata%nt
        do j = 1, gdata%nloci
            print "(F5.1)", gdata%X(i, j)
        end do
    end do

end program plink_test
